enable the use of iaas-specific kernels

rthill91 opened the pull request:
Use `UBUNTU_IAAS_KERNEL` environment variable to identify the use of, and label for, iaas specific kernels (e.g. `aws`)

This somewhat replicates how fips was handled in the bionic branch, but with an attempt to be more generic.

rthill91 committed:
Enable the use of iaas-specific kernels

toabctl commented on bosh-stemcell/lib/bosh/stemcell/builder_options.rb:
```diff
@@ -63,6 +63,7 @@ def environment_variables
         'UBUNTU_ISO' => environment['UBUNTU_ISO'],
         'UBUNTU_MIRROR' => environment['UBUNTU_MIRROR'],
         'UBUNTU_ADVANTAGE_TOKEN' => environment['UBUNTU_ADVANTAGE_TOKEN'],
+        'IAAS_KERNEL' => environment['IAAS_KERNEL'],
```
nit: should this be `UBUNTU_IAAS_KERNEL` given that this is Ubuntu specific?

toabctl replied:
would be nice to check that this var can only be set to `aws` currently.

rthill91 replied:
renamed in https://github.com/cloudfoundry/bosh-linux-stemcell-builder/pull/309/commits/fed91af0a965655cfb6efa15085874f34de66863 and added a check for the value in https://github.com/cloudfoundry/bosh-linux-stemcell-builder/pull/309/commits/21485472676b68adf8bb711443abff0c92824d1c

toabctl commented on bosh-stemcell/spec/stemcells/fips_spec.rb:
```diff
@@ -2,7 +2,11 @@
   context 'installed by system_kernel' do
-    describe package('linux-image-fips') do
+    iaas = ENV['IAAS_KERNEL'] || ''
+    unless iaas.empty?
```
is this check for empty required? isn't a check for `ENV['IAAS_KERNEL']` enough? at least that's what's done in other places in this PR.

rthill91 replied:
I need to append that hyphen if `IAAS_KERNEL` isn't empty otherwise I want an empty string.

I could drop this in favor of the following oneliner
```rb
iaas = ("#{ENV['IAAS_KERNEL']}-" if ENV['IAAS_KERNEL']) || ''
```

ruby is also not my forte, so maybe I'm missing a more obvious way to do this.

toabctl replied:
thx (looks like I can't "resolve" this thread). Can somebody mark it as resolved please?

rthill91 committed:
rename IAAS_KERNEL -> UBUNTU_IAAS_KERNEL

rthill91 committed:
verify that UBUNTU_IAAS_KERNEL value is valid

rkoster commented:
@toabctl Is this PR ready for review? asking because there currently still is active discussion.

rkoster commented on stemcell_builder/stages/system_fips_kernel/apply.sh:
```diff
@@ -14,6 +14,10 @@ add_on_exit "umount $chroot/sys"
+if [ ! -z "$UBUNTU_IAAS_KERNEL" ]; then
+    kernel=linux-$UBUNTU_IAAS_KERNEL-fips
+fi
+pkg_mgr install fips-initramfs $kernel
```
Do we still need to specify the `cloud_type` if we install iaas specific kernel here anyway? I believe that code was only needed in the past when we relied on ua attach to install the kernel. But now that ua attach [runs without apt](https://github.com/cloudfoundry/bosh-linux-stemcell-builder/pull/309/files#diff-4741572db64c272a4d3237f9fbaeda98a4788cee27e1e1a1024c31bab6752b25R53-R62), this should no longer be needed.

rthill91 replied:
I think you're correct. We should be able to drop the client config overwrite altogether.

rkoster commented on bosh-stemcell/lib/bosh/stemcell/builder_options.rb:
```diff
@@ -63,6 +63,7 @@ def environment_variables
         'UBUNTU_ISO' => environment['UBUNTU_ISO'],
         'UBUNTU_MIRROR' => environment['UBUNTU_MIRROR'],
         'UBUNTU_ADVANTAGE_TOKEN' => environment['UBUNTU_ADVANTAGE_TOKEN'],
+        'UBUNTU_IAAS_KERNEL' => environment['UBUNTU_IAAS_KERNEL'],
```
To make this future proof I would like to propose change the variable name to: `UBUNTU_USE_IAAS_KERNEL` when set would install the iaas specific kernel for the given `$stemcell_infrastructure` (this variable is already available and gets used for example [here](https://github.com/cloudfoundry/bosh-linux-stemcell-builder/blob/ubuntu-jammy/master/stemcell_builder/stages/image_install_grub/apply.sh#L78C9-L85)).

Even more correct would be to use `UBUNTU_FIPS_USE_IAAS_KERNEL` because that's the current scope of the variable.

rkoster replied:
To be able to use the `infrastructure.name` in the specs you would have to extend the rspec invocation [here](https://github.com/cloudfoundry/bosh-linux-stemcell-builder/blob/ubuntu-jammy/master/bosh-stemcell/lib/bosh/stemcell/build_environment.rb#L52-L56) to something like:
```ruby
      cmd = [
        "cd #{STEMCELL_SPECS_DIR};",
        "STEMCELL_IMAGE=#{image_file_path}",
        "STEMCELL_WORKDIR=#{work_path}",
        "STEMCELL_INFRASTRUCTURE=#{infrastructure.name}",  # ← added
        "OS_NAME=#{operating_system.name}",
        "OS_VERSION=#{operating_system.version}",
        "CANDIDATE_BUILD_NUMBER=#{@version}",
        "bundle exec rspec -fd#{exclude_exclusions}",
        "spec/os_image/#{operating_system_spec_name}_spec.rb",
        "spec/stemcells/#{operating_system_spec_name}_spec.rb",
        'spec/stemcells/go_agent_spec.rb',
        "spec/stemcells/#{infrastructure.name}_spec.rb",
        'spec/stemcells/stig_spec.rb',
        'spec/stemcells/cis_spec.rb'
      ]
```

rkoster commented on bosh-stemcell/spec/stemcells/fips_spec.rb:
```diff
@@ -2,7 +2,11 @@
+    end
+    describe package("linux-image-#{iaas}fips") do
       it { should be_installed }
     end
```
I would use something like:
```ruby
let(:infrastructure) { ENV['STEMCELL_INFRASTRUCTURE'] }
let(:use_iaas_kernel) { ENV['UBUNTU_FIPS_USE_IAAS_KERNEL'] }
describe package(use_iaas_kernel ? "linux-image-#{infrastructure}-fips" : "linux-image-fips") do
  it { should be_installed }
end
```

It could be a one-liner but given the environment variable name lengths, something like the above would be more readable.

rthill91 replied:
I'm getting errors trying to use those variables inside describe. Specifically
> `use_iaas_kernel` is not available on an example group (e.g. a `describe` or `context` block). It is only available from within individual examples (e.g. `it` blocks) or from constructs that run in the scope of an example (e.g. `before`, `let`, etc).

Is there a scope concern keeping these as instance variables defined within this test? e.g.

```ruby
context 'installed by system_kernel' do
  infrastructure = ENV['STEMCELL_INFRASTRUCTURE']
  use_iaas_kernel = ENV['UBUNTU_FIPS_USE_IAAS_KERNEL']
  
  describe package(use_iaas_kernel ? "linux-image-#{infrastructure}-fips" : "linux-image-fips") do
    it { should be_installed }
  end
  ...
end
```

rkoster replied:
Yeah, using normal variables seems fine. I did not test my example, was just meant as an illustration.

rkoster reviewed (changes requested):
Would like this PR to refactored a bit so that the target infrastructure does not have to be specified twice on rake invocation, by making the introduced variable only toggle and not use it for passing down the target infrastructure into the tests.

rthill91 committed:
remove unneeded client config overwrites

Specifying the use of an IAAS specific kernel via environment variables
removes the need for the client config updates.

rthill91 committed:
change UBUNTU_IAAS_KERNEL to UBUNTU_FIPS_USE_IAAS_KERNEL

change the iaas kernel env var to work as a toggle for using the
infrastructure.name for iaas_specific work as opposed to definining it
directly

rkoster commented on bosh-stemcell/spec/stemcells/fips_spec.rb:
```diff
@@ -49,12 +55,35 @@
     let(:dpkg_list_azure_ubuntu) { File.readlines(spec_asset('dpkg-list-ubuntu-jammy-azure-additions.txt')).map(&:chop) }
     let(:dpkg_list_cloudstack_ubuntu) { File.readlines(spec_asset('dpkg-list-ubuntu-jammy-cloudstack-additions.txt')).map(&:chop) }
     let(:dpkg_list_softlayer_ubuntu) { File.readlines(spec_asset('dpkg-list-ubuntu-jammy-softlayer-additions.txt')).map(&:chop) }
 
```
I would move infrastructure and use_iaas_kernel here, so you do not have to duplicate them in nested describes.
```
      let(:infrastructure) { ENV['STEMCELL_INFRASTRUCTURE'] }
      let(:use_iaas_kernel) { ENV['UBUNTU_FIPS_USE_IAAS_KERNEL'] }
```

rkoster commented on bosh-stemcell/spec/stemcells/fips_spec.rb:
```diff
@@ -49,12 +55,35 @@
+      let(:use_iaas_kernel) { ENV['UBUNTU_FIPS_USE_IAAS_KERNEL'] }
+
+      it 'contains only the base set of packages plus aws-specific kernel packages' do
+        skip "Test skipped due to non-aws kernel" if (use_iaas_kernel && infrastructure != 'aws')
```
Could this line be changed to:
```
skip "Test skipped due to generic kernel" unless use_iaas_kernel
```
The infrastructure check should not be needed because the excludes.

rkoster commented on bosh-stemcell/spec/stemcells/fips_spec.rb:
```diff
@@ -37,7 +39,11 @@
       it { should be_file }
       its(:content) { should match %r{linux\t/boot/vmlinuz-\S+-fips root=UUID=\S* ro } }
-      its(:content) { should match %r{initrd\t/boot/initrd.img-\S+-fips} }
+      if ENV.key?("UBUNTU_FIPS_USE_IAAS_KERNEL")
```
Have you checked if the else can be reached? It could be that this environment variable always will be set, just to an empty string. Because of: https://github.com/cloudfoundry/bosh-linux-stemcell-builder/blob/04e622ab133a1d88a4a8b4b4685498925372c33e/bosh-stemcell/lib/bosh/stemcell/builder_options.rb#L66

rthill91 replied:
sorry I had to put this down for a couple weeks I've added a check for empty string as well as I think you're right about all the envvars being 'set' regardless

rthill91 committed:
ensure iaas_kernel envvars are set and not empty in testing

rkoster commented on bosh-stemcell/spec/stemcells/fips_spec.rb:
```diff
@@ -63,7 +91,10 @@
       exclude_on_azure: true,
       exclude_on_softlayer: true,
     } do
+      let(:use_iaas_kernel) { ENV['UBUNTU_FIPS_USE_IAAS_KERNEL'] }
```
This line should not be needed because `use_iaas_kernel` is already defined here: https://github.com/cloudfoundry/bosh-linux-stemcell-builder/pull/309/files#diff-f18a6346d0ba4575fc99197220088b78b3e21a9696f0c5949166c0bb52463744R65

rkoster commented on bosh-stemcell/spec/stemcells/fips_spec.rb:
```diff
@@ -49,11 +55,33 @@
     let(:dpkg_list_cloudstack_ubuntu) { File.readlines(spec_asset('dpkg-list-ubuntu-jammy-cloudstack-additions.txt')).map(&:chop) }
     let(:dpkg_list_softlayer_ubuntu) { File.readlines(spec_asset('dpkg-list-ubuntu-jammy-softlayer-additions.txt')).map(&:chop) }
+    let(:infrastructure) { ENV['STEMCELL_INFRASTRUCTURE'] }
+    let(:use_iaas_kernel) { ENV['UBUNTU_FIPS_USE_IAAS_KERNEL'] and ENV['UBUNTU_FIPS_USE_IAAS_KERNEL'] != "" }
```
This could be written as:
```ruby
let(:use_iaas_kernel) { ENV.fetch('UBUNTU_FIPS_USE_IAAS_KERNEL', 'false') == 'true' }
```

rkoster commented on bosh-stemcell/spec/stemcells/fips_spec.rb:
```diff
@@ -2,7 +2,9 @@
   context 'installed by system_kernel' do
-    describe package('linux-image-fips') do
+    infrastructure = ENV['STEMCELL_INFRASTRUCTURE']
+    use_iaas_kernel = ENV['UBUNTU_FIPS_USE_IAAS_KERNEL'] and ENV['UBUNTU_FIPS_USE_IAAS_KERNEL'] != ""
```
```
use_iaas_kernel = ENV.fetch('UBUNTU_FIPS_USE_IAAS_KERNEL', 'false') == 'true'
```

rkoster reviewed (changes requested):
Almost there, just some small nitpicks.

rkoster commented on bosh-stemcell/spec/stemcells/fips_spec.rb:
```diff
@@ -3,8 +3,8 @@
-    use_iaas_kernel = ENV['UBUNTU_FIPS_USE_IAAS_KERNEL']
-    describe package(use_iaas_kernel ? "linux-image-#{infrastructure}-fips" : "linux-image-fips") do
+    use_iaas_kernel = ENV.fetch('UBUNTU_FIPS_USE_IAAS_KERNEL', 'false') != 'true'
+    describe package(use_iaas_kernel != "" ? "linux-image-#{infrastructure}-fips" : "linux-image-fips") do
```
`use_iaas_kernel` already is a boolean, no need to check against empty string

rkoster commented on bosh-stemcell/spec/stemcells/fips_spec.rb:
```diff
@@ -39,7 +39,7 @@
       it { should be_file }
       its(:content) { should match %r{linux\t/boot/vmlinuz-\S+-fips root=UUID=\S* ro } }
-      if ENV.key?("UBUNTU_FIPS_USE_IAAS_KERNEL")
+      if ENV.key?("UBUNTU_FIPS_USE_IAAS_KERNEL") and ENV["UBUNTU_FIPS_USE_IAAS_KERNEL"] != ""
```
The `use_iaas_kernel` let from further down [below](https://github.com/cloudfoundry/bosh-linux-stemcell-builder/pull/309/commits/6582e8d911973925a0fd354f2fa5f44adc4cfe6a#diff-f18a6346d0ba4575fc99197220088b78b3e21a9696f0c5949166c0bb52463744R65) should be moved up so that it can be reused here.

rkoster reviewed (approved):
LGTM :shipit:

ramonskie reviewed (approved):
looks good

rkoster commented:
Thanks! @rthill91
